File: frontend/lib/features/habits/domain/habit_repository.dart
import 'habit.dart';
import 'habit_entry.dart';

abstract class HabitRepository {
  Future<List<Habit>> getHabits();

  Future<Habit> createHabit(Habit habit);

  Future<HabitEntry> addHabitEntry(String habitId, HabitEntry entry);

  Future<List<HabitEntry>> getHabitEntries(String habitId, {DateTime? startDate, DateTime? endDate});

  Future<void> deleteHabit(String habitId);

  Future<void> toggleHabitCheck(String habitId, bool completed);
}

File: app/src/main/java/com/example/cozytrack/domain/repository/HabitRepository.kt
package com.example.cozytrack.domain.repository

import com.example.cozytrack.core.network.ApiResult
import com.example.cozytrack.domain.model.Frequency
import com.example.cozytrack.domain.model.Habit
import com.example.cozytrack.domain.model.HabitEntry
import com.example.cozytrack.domain.model.HabitProgress
import com.example.cozytrack.domain.model.ServerClock

interface HabitRepository {
    suspend fun getServerClock(): ApiResult<ServerClock>
    suspend fun getHabits(includeArchived: Boolean = false): ApiResult<List<Habit>>
    suspend fun getHabit(habitId: String): ApiResult<Habit>
    suspend fun createHabit(title: String, frequency: Frequency, startDate: String): ApiResult<String>
    suspend fun updateHabit(
        habitId: String,
        title: String? = null,
        frequency: Frequency? = null,
        startDate: String? = null,
        isArchived: Boolean? = null
    ): ApiResult<Unit>
    suspend fun deleteHabit(habitId: String): ApiResult<Unit>
    suspend fun checkHabit(habitId: String, completed: Boolean): ApiResult<Unit>
    suspend fun getEntriesForDay(checkDate: String): ApiResult<List<HabitEntry>>
    suspend fun getHabitEntries(
        habitId: String,
        startDate: String? = null,
        endDate: String? = null
    ): ApiResult<List<HabitEntry>>
    suspend fun getHabitProgress(
        habitId: String,
        startDate: String? = null,
        endDate: String? = null
    ): ApiResult<HabitProgress>
}
